File: android/helper/src/main/java/org/dweb_browser/helper/ChangeableMap.kt
package org.dweb_browser.helper

import kotlinx.serialization.Serializable
import kotlin.coroutines.CoroutineContext

@Serializable
data class ChangeState<K>(val adds: Set<K>, val updates: Set<K>, val removes: Set<K>);

class ChangeableMap<K, V>(context: CoroutineContext = ioAsyncExceptionHandler) :
  LinkedHashMap<K, V>() {
  data class Changes<K, V>(
    val origin: ChangeableMap<K, V>, val adds: Set<K>, val updates: Set<K>, val removes: Set<K>
  )

  private val changeable = Changeable(Changes(this, setOf(), setOf(), setOf()), context)
  fun setContext(context: CoroutineContext) {
    changeable.context = context
  }

  val onChange = changeable.onChange
  suspend fun emitChange() = changeable.emitChange()
  fun emitChangeBackground(
    adds: Set<K> = setOf(),
    updates: Set<K> = setOf(),
    removes: Set<K> = setOf()
  ) {
    changeable.emitChangeBackground(Changes(this, adds, updates, removes))
  }

  override fun clear() {
    if (keys.size > 0) {
      val removes = keys.toSet()
      super.clear().also { emitChangeBackground(removes) }
    }
  }

  override fun put(key: K, value: V) = super.put(key, value).also {
    if (it != null) {// 有前置的值，那么就触发 update
      emitChangeBackground(updates = setOf(key))
    } else {// 没有前置的值，那么就触发 add
      emitChangeBackground(adds = setOf(key))
    }
  }

  private operator fun set(key: K, value: V) {
    this.put(key, value)
  }

  override fun putAll(from: Map<out K, V>) {
    val updates = mutableSetOf<K>()
    val adds = mutableSetOf<K>()
    for (key in from.keys) {
      if (contains(key)) {
        updates.add(key)
      } else {
        adds.add(key)
      }
    }
    super.putAll(from).also { emitChangeBackground(updates, adds) }
  }

  override fun remove(key: K) =
    super.remove(key)?.also { emitChangeBackground(removes = setOf(key)) }


  /** 重置 清空所有的事件监听，清空所有的数据
   * 注意，这里不会触发任何事件，如果有需要，请使用 clear ，然后再 reset
   */
  fun reset() {
    changeable.signal.clear()
    this.clear()
  }
}

File: next/dweb-browser/Helper/src/ChangeableMap.cs
using System.Collections.Concurrent;

namespace DwebBrowser.Helper;

public record Changes<K, V>(ChangeableMap<K, V> Origin, HashSet<K> Adds, HashSet<K> Updates, HashSet<K> Removes);

public record ChangeState<K>(HashSet<K> Adds, HashSet<K> Updates, HashSet<K> Removes);

public class ChangeableMap<K, V> : ConcurrentDictionary<K, V> where K : notnull
{
    private readonly LazyBox<Changeable<Changes<K, V>>> LazyChangeable = new();
    private Changeable<Changes<K, V>> Changeable =>
        LazyChangeable.GetOrPut(() => new Changeable<Changes<K, V>>(new Changes<K, V>(this, new(), new(), new())));

    public void OnChangeAdd(Signal<Changes<K, V>> cb)
    {
        Changeable.Listener.OnListener += cb;
    }

    public void OnChangeRemove(Signal<Changes<K, V>> cb)
    {
        Changeable.Listener.OnListener -= cb;
    }

    public void OnChangeEmit()
    {
        Changeable.EmitChange();
    }

    public void OnChangeBackgroundEmit(HashSet<K>? adds = null, HashSet<K>? updates = null, HashSet<K>? removes = null)
    {
        adds ??= new();
        updates ??= new();
        removes ??= new();

        Changeable.EmitChangeBackground(new Changes<K, V>(this, adds, updates, removes));
    }

    public new bool TryAdd(K key, V value)
    {
        if (Keys.Contains(key))
        {
            return TryUpdate(key, value, Get(key)).Also(it =>
            {
                if (it)
                {
                    OnChangeBackgroundEmit(updates: new HashSet<K> { key });
                }
            });
        }
        else
        {
            return base.TryAdd(key, value).Also(it =>
            {
                if (it)
                {
                    OnChangeBackgroundEmit(adds: new HashSet<K> { key });
                }
            });
        }
    }

    public bool Set(K key, V value)
    {
        return TryAdd(key, value);
    }

    public V? Get(K key)
    {
        if (TryGetValue(key, out var value))
        {
            return value;
        }

        return default;
    }

    public V? Remove(K key)
    {
        TryRemove(key, out var value);

        return value;
    }

    public bool TryRemove(K key)
    {
        return TryRemove(key, out _);
    }

    public new bool TryRemove(K key, out V value)
    {
        var _bool = base.TryRemove(key, out var baseValue);

        if (_bool)
        {
            value = baseValue;
            OnChangeBackgroundEmit(removes: new HashSet<K> { key });
        }
        else
        {
            value = default;
        }

        return _bool;
    }

    public new void Clear()
    {
        if (Keys.Count > 0)
        {
            var removes = Keys.ToHashSet();
            base.Clear();

            OnChangeBackgroundEmit(removes: removes);
        }
    }

    /// <summary>
    /// 重置 清空所有的事件监听，清空所有的数据
    ///
    /// 注意，这里不会触发任何事件，如果有需要，请使用 clear ，然后再 reset
    /// </summary>
    public void Reset()
    {
        Changeable.Listener.Clear();
        Clear();
    }
}
